server/game/cards/01_SOR/upgrades/JediLightsaber.ts

```diff
@@ -0,0 +1,41 @@
+import AbilityHelper from '../../../AbilityHelper';
+import { Attack } from '../../../core/attack/Attack';
+import { Card } from '../../../core/card/Card';
+import { UpgradeCard } from '../../../core/card/UpgradeCard';
+import { Trait } from '../../../core/Constants';
+import Player from '../../../core/Player';
+
+export default class JediLightsaber extends UpgradeCard {
+    protected override getImplementationId() {
+        return {
+            id: '8495694166',
+            internalName: 'jedi-lightsaber',
+        };
+    }
+
+    public override canAttach(targetCard: Card, controller: Player = this.controller): boolean {
+        if (targetCard.hasSomeTrait(Trait.Vehicle)) {
+            return false;
+        }
+
+        return super.canAttach(targetCard, controller);
+    }
+
+    public override setupCardAbilities() {
+        this.addGainTriggeredAbilityTargetingAttached({ title: 'Give the defender -2/-2 for this phase',
+            when: { onAttackDeclared: (event, context) => event.attack.attacker === context.source },
+
+            // need to check if the target is a base - if so, don't apply the stat modifier effect
```

Comment:
Can we mess with canAffect() or maybe StaticOngoingEffect.canBeApplied() so that the effect just isn't applied if it's invalid for the target? I wanted basically that(except for targeting) for an ability that gave sentinel, but couldn't figure it out and made it a TODO. What do you think?

Reply:
Yeah, generally speaking it would be useful if an effect could declare default target conditions the way a GameSystem does.

It would be a pretty big change to the way effects are set up though, and the whole effect system is badly in need of an overhaul as is. So I'm kind of inclined to postpone any big improvements on the system until after launch, unless it starts to become a really big issue

Reply:
Actually, I'll poke around at it and evaluate how bad it would actually be. There's definitely stuff that would be easier if it was changed